hca/dss/__init__.py

```diff
@@ -107,87 +111,181 @@ def download(self, bundle_uuid, replica, version="", dest_name="",
+                            logger.info("%s", "File {}: Resuming at {}.".format(
+                                file_uuid, server_start))
+                        elif consume_bytes > 0:
+                            logger.info("%s", "File {}: Resuming at {}. Dropping {} bytes to match".format(
+                                file_uuid, server_start, consume_bytes))
+
+                            while consume_bytes > 0:
+                                bytes_to_read = min(consume_bytes, 1024*1024)
+                                content = response.iter_content(chunk_size=bytes_to_read)
+                                chunk = next(content)
+                                if chunk:
+                                    consume_bytes -= len(chunk)
+
+                        for chunk in response.iter_content(chunk_size=1024*1024):
+                            if chunk:
+                                fh.write(chunk)
+                                hasher.update(chunk)
+                                retries_left = min(retries_left + 1, num_retries)
+                                delay = max(delay / 2, min_delay_seconds)
+                        break
+                    finally:
+                        response.close()
+                except (ChunkedEncodingError, ConnectionError, ReadTimeout):
+                    if retries_left > 0:
+                        logger.info("%s", "File {}: GET FAILED. Attempting to resume.".format(
+                            file_uuid, dest_path))
+                        time.sleep(delay)
+                        delay *= 2
+                        retries_left -= 1
+                        continue
+                    raise
+
+            if hasher.hexdigest().lower() != file_sha256.lower():
+                os.remove(dest_path)
+                logger.error("%s", "File {}: GET FAILED. Checksum mismatch.".format(file_uuid))
                 raise ValueError("Expected sha256 {} Received sha256 {}".format(
-                    file_["sha256"].lower(), hasher.hexdigest().lower()))
+                    file_sha256.lower(), hasher.hexdigest().lower()))
+            else:
+                logger.info("%s", "File {}: GET SUCCEEDED. Stored at {}.".format(file_uuid, dest_path))
+
+    @classmethod
+    def _file_path(cls, checksum):
+        """
+        returns a file's relative local path based on the nesting parameters and the files hash
+        :param checksum: a string checksum
+        :return: relative Path object
+        """
+        assert(sum(cls.DIRECTORY_NAME_LENGTHS) <= len(checksum))
+        checksum_index = 0
+        path_pieces = []
+        for prefix_length in cls.DIRECTORY_NAME_LENGTHS:
+            path_pieces.append(checksum[checksum_index:(checksum_index + prefix_length)])
+            checksum_index += prefix_length
+        path_pieces.append(checksum)
+        return os.path.join(*path_pieces)
+
+    @classmethod
+    def _parse_manifest(cls, manifest):
```

Comment:
The way I wrote this, the entire file is read into memory at once. Will we ever encounter a manifest large enough for this to be a problem?

Reply:
I doubt it.